### PR TITLE
Tool to search and replace bookie ids in ledger metadata

### PR DESCRIPTION
**Pull request by ivankelly**
To use:
```
bin/bkctl bookieid searchreplace --from <from> --to <to>
```

To be used in cases where the DNS name of the bookie has to change, and you don't want all the data to have to be moved by autorecovery.

### CODE REVIEW

File: tools/ledger/src/main/java/org/apache/bookkeeper/tools/cli/commands/bookieid/SearchReplaceBookieIdCommand.java [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+        super(CliSpec.<Flags>newBuilder()
+            .withName(NAME)
+            .withDescription(DESC)
+            .withFlags(flags)
+            .build());
+    }
+
+    @Override
+    protected void run(BookKeeper bk, Flags flags) throws Exception {
+        BookKeeperAdmin admin = new BookKeeperAdmin((org.apache.bookkeeper.client.BookKeeper) bk);
```
Comment (sijie) [inline tools/ledger/src/main/java/org/apache/bookkeeper/tools/cli/commands/bookieid/SearchReplaceBookieIdCommand.java]: might be worth closing admin at the end?

File: tools/ledger/src/main/java/org/apache/bookkeeper/tools/cli/commands/bookieid/SearchReplaceBookieIdCommand.java [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+                                    } else {
+                                        return a;
+                                    }
+                                });
+                            builder.replaceEnsembleEntry(e.getKey(), ensemble);
+                        });
+                LedgerMetadata newMeta = builder.build();
+                if (flags.verbose) {
+                    System.out.println("Replacing ledger " + lid + " metadata ...");
+                    System.out.println(md.getValue().toString());
```
Comment (sijie) [inline tools/ledger/src/main/java/org/apache/bookkeeper/tools/cli/commands/bookieid/SearchReplaceBookieIdCommand.java]: use `toSafeString()`?

File: tools/ledger/src/main/java/org/apache/bookkeeper/tools/cli/commands/bookieid/SearchReplaceBookieIdCommand.java [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+                    System.out.println("with ...");
+                    System.out.println(newMeta.toString());
```
Comment (sijie) [inline tools/ledger/src/main/java/org/apache/bookkeeper/tools/cli/commands/bookieid/SearchReplaceBookieIdCommand.java]: use `toSafeString()`?

**Comment by jvrao** [comment]
@ivankelly  @sijie  what is the use case of this utility? can you please give little more info ?

**Comment by sijie** [comment]
@jvrao @eolivelli I thought it is related to #634 , but after looking into BookieShell, there is already a comment `updateledgers`. so I am also a bit confused about this PR. @ivankelly can you explain more about the context?

**Comment by ivankelly** [comment]
@jvrao @sijie @eolivelli The usecase is that a client needed to update the DNS addresses of their bookies. The data already existed on the bookie, but the DNS and advertised address was changing (I had to update the cookies manually with sed, it was a stateful set in k8s). I wanted to avoid autorecovery trying to recovery every existing ledger, since the data was still where it should have been.

```updateledgers``` does exist, but it's very limited. It only allows changing from ip to hostname or vice versa based on what is in the configuration file. Even on baremetal, this is awkward to use, but on k8s it's completely non-viable.

**Comment by ivankelly** [comment]
run pr validation

**Comment by ivankelly** [comment]
> Can you also provide a command to update the cookies (as you did with 'sed')?

I don't have the cycles to do so right now (with good testing) but i'll put it on the backlog.

**Comment by sijie** [comment]
@ivankelly 

there is one checkstyle issue, can you fix it?

```
[ERROR] /home/jenkins/jenkins-slave/workspace/bookkeeper_precommit_pullrequest_validation/tools/ledger/src/main/java/org/apache/bookkeeper/tools/cli/commands/bookieid/SearchReplaceBookieIdCommand.java:92: Line is longer than 120 characters (found 121). [LineLength]
```

**Comment by jiazhai** [comment]
@ivankelly 
Is it good to add your use case into the PR description? 
```
The usecase is that a client needed to update the DNS addresses of their bookies. The data already existed on the bookie, but the DNS and advertised address was changing (I had to update the cookies manually with sed, it was a stateful set in k8s). I wanted to avoid autorecovery trying to recovery every existing ledger, since the data was still where it should have been.

updateledgers does exist, but it's very limited. It only allows changing from ip to hostname or vice versa based on what is in the configuration file. Even on baremetal, this is awkward to use, but on k8s it's completely non-viable.
```

**Comment by sijie** [comment]
run pr validation

**Pull request merged**